sycl/include/CL/sycl/detail/cg.hpp

```diff
@@ -36,18 +36,66 @@ namespace sycl {
 
 // Forward declarations
 class queue;
-namespace detail {
-class queue_impl;
-} // namespace detail
 
 namespace detail {
 
+// Used to represent a type of an extended member
+enum class ExtendedMembersType: unsigned int {
+  HANDLER_KERNEL_BUNDLE = 0,
+};
+
+// Holds a pointer to an object of an arbitrary type and an ID value which
+// should be used to understand what type pointer points to.
+// Used as to extend handler class without introducing new class members which
+// would change handler layout.
+struct ExtendedMemberT {
+  ExtendedMembersType MType;
+  std::shared_ptr<void> MData;
+};
+
+static std::shared_ptr<std::vector<ExtendedMemberT>>
+convertToExtendedMembers(const std::shared_ptr<const void> &SPtr) {
+  return std::const_pointer_cast<std::vector<ExtendedMemberT>>(
+      std::static_pointer_cast<const std::vector<ExtendedMemberT>>(SPtr));
+}
+
 class stream_impl;
+class queue_impl;
+class kernel_bundle_impl;
+
+// The version of CG type is encoded in the higher byte of the value:
+//
+// 0x00000001 - CG type KERNEL version 0
+// 0x01000001 - CG type KERNEL version 1
+//   /\
+//   ||
+// The byte describes the version
+//
+// The constant is used to shift left CG type value to access it's version
+constexpr unsigned int ShiftBitsForVersion = 24;
+
+// Constructs versioned type
+constexpr unsigned int getVersionedCGType(unsigned int Type,
+                                          unsigned char Version) {
+  return Type | ((unsigned int)Version << ShiftBitsForVersion);
+}
+
+// Returns the version encoded to the type
+constexpr unsigned char getCGTypeVersion(unsigned int Type) {
+  return Type >> ShiftBitsForVersion;
+}
+
 /// Base class for all types of command groups.
 class CG {
 public:
+
+  enum class CG_VERSION: unsigned char {
```

Comment:
Nit: 128 versions should probably be enough, but this discords with other enums. If there is perf/size consideration behind unsigned char selection - please add comment.

Reply:
Added comment.